Fix: Support exceptions in no-multi-spaces (fixes #1755)

btmills opened the pull request.

nzakas commented:
Code looks good, though I think this is a breaking change. By specifying no exceptions by default, that means anyone using both this and `key-spacing` will need to update their config. Maybe instead of using an array of exceptions it should be an object? Like:

```
no-multi-spaces: [2, {
    Property: true
}]
```

Or some such thing so we can set `Property` to be ignored by default?

Also, I'm a bit concerned that a problem in an object literal method will be missed, such as:

```js
var foo = {
    bar: function() {
        return 1      + 2;
    }
};
```

Can you double-check what happens in that case?

btmills commented:
Is this breaking? The rule conflict was introduced in 8144f13. As it is right now, this PR doesn't change any behavior; one must enable an exception to have any effect.

```js
// test.js
/* eslint no-multi-spaces: [2, { exceptions: ["Property"] }] */
/* eslint key-spacing: [2, { align: "value" }] */
var obj = {
	first:  "first",
	second: "second"
};
```

On `master`:

```sh
$ eslint --reset --no-eslintrc test.js

test.js
  5:9  error  Multiple spaces found before '"first"'  no-multi-spaces

✖ 1 problem (1 error, 0 warnings)
```

On `issue1756` it exits without any warnings.

However, I think excepting `Property` by default might circumvent the question entirely. I'll do that.

btmills commented:
Updated. `Property` is now excepted by default, `exceptions` is an object instead of an array, and I added a test case specifically for the object literal method you gave. Since the exception only applied to the immediate parent, problems nested within an excepted node still get picked up.

nzakas commented on lib/rules/no-multi-spaces.js:
```diff
@@ -12,7 +14,17 @@
+        Object.keys(options.exceptions).forEach(function (key) {
+            exceptions[key] = options.exceptions[key];
+        });
+        exceptionCount = Object.keys(exceptions).length;
```
I don't think `exceptionCount` is needed. There's always at least one entry in exceptions, so this is always truthy.

btmills replied:
Ah, oops. The eventual intention was to `delete` if the exception is `false`. The goal is to avoid the performance cost of checking parents if there are no exceptions. This will be fixed in a moment.

btmills replied:
Done.

nzakas replied:
Did you forget to commit the changes? Still looks the same to me.

btmills replied:
Hmm... is the page cached? This portion is showing as an outdated diff for me. If I go to the Files changed tab this section is an if/else, setting in the consequent and deleting in the alternate.

btmills replied:
The new commit hash is 8252d28.

nzakas replied:
I'm still seeing `exceptionCount` being used in the same way on that commit. I'm expecting to see it removed.

btmills replied:
The goal is to check the type of the parent node only if there are exceptions. If there are no exceptions, it doesn't matter what the parent is, so there's no reason to traverse down the AST to find it. So it won't go away, I just need to make sure it's actually deleting keys rather than setting them to false. That way if someone overrides and removes `Property`, `exceptionCount` will be 0, and no traversal will occur.

nzakas replied:
Hmmm, still confusing. Maybe rename to `hasExceptions` and use `> 1` to convert the count into a Boolean?

btmills replied:
That's much more understandable. Done.

nzakas commented on lib/rules/no-multi-spaces.js:
```diff
@@ -65,9 +78,15 @@ module.exports = function(context) {
-                        context.report(token, token.loc.start,
-                            "Multiple spaces found before '{{value}}'.",
-                            { value: token.value });
+                        if (exceptionCount) {
```
Looks like this always evaluates to true. You may be able to remove this check.

nzakas commented:
Lgtm.
